rfcs/0074-refine-s3-source-in-lazy-way.md

```diff
@@ -0,0 +1,87 @@
+needs to transmit the entire mapping relationship to every `Source Executor`. This can lead to excessive transmission overhead, 
+prolonging the recovery time and potentially stalling the entire recovery process. 
+As time progresses, the number of files that have been read increases. However, we still need to send the entire mapping 
+relationship every time, which means that most of the transmitted data pertains to files that have already been read, 
+rather than files that need to be read.
+
+An obvious optimization strategy is to avoid sending files that have already been read. We can maintain a list of 
+finished files in `Source Manager` and clear the related status in the table fragment. 
+When transmitting the mapping relationship, only the unfinished files should be sent.
+
+To achieve this goal, a proper mechanism is needed to recycle the list of finished files. 
+The good news is that the barrier can assist in this process. When each checkpoint arrives, the source executor can 
+write the filenames of the finished files from the state cache into the barrier. 
+Eventually, through barrier collection, this information is transmitted to the Source Manager, completing the process. 
+The changes involved here include adding a new field or Mutation called `FinishedFiles` in the barrier 
+and introducing a new state `finishedFiles` in `Source Manager`. 
+This status will record the filenames of the finished or deleted files and needs to be persisted.
+Additionally, the `collect_barrier` needs to be modified to support retrieving `FinishedFiles` from the finished barriers.
+
+Another optimization is that since each barrier will carry at most one mutation, the content of Add/SplitChange 
+can be extracted when flowing through the `source executor`, reducing the overhead of subsequent transmissions.
+
+### Control the Number of Ongoing Files
+
+In the previously mentioned solution, we focused on how to avoid sending too many files during the reading process. 
+However, if we send all the files in the s3 bucket at once, we still face issues with an overly large barrier and prolonged list objects times. 
+Therefore, we can attempt to control the number of files sent each time, allowing us to manage the overall load on the cluster.
+
+From our introduction to the S3 Source, we know that the Source Manager can determine the total number of files 
+and the number of files that have been completed. With this information, we can ascertain the number of files currently being read, aka. ongoing files. 
+Taking into account the limitations of the S3 API, I propose a solution based on reading file list pagination 
+according to the number of files being read.
+
+1. Define the current page `cur_page`, initialized to 1 (indicating the max page per `list objects` request), 
+and the trigger to read the next page's file count `MIN_ONGOING_FILE_NUM`. This value should be less than 1000.
```

Comment:
It is ok to add an option here. 
I propose 1000 here because the maximum number of items is 1000 per page in list-objects request.
The proposal I put forward is to control on a per-page basis, so I believe the control level should be an integer multiple of the single-page capacity.